src/app/fdctl/run/tiles/fd_gui.c

```diff
@@ -0,0 +1,342 @@
+  .max_ws_send_frame_cnt = FD_HTTP_SERVER_GUI_MAX_WS_SEND_FRAME_CNT,
+  .outgoing_buffer_sz    = FD_HTTP_SERVER_GUI_OUTGOING_BUFFER_SZ,
+};
+
+typedef struct {
+  fd_topo_t * topo;
+
+  fd_gui_t * gui;
+  uchar      buf[ 8UL+40200UL*(58UL+12UL*34UL) ] __attribute__((aligned(8)));
+
+  fd_http_server_t * gui_server;
+
+  char          version_string[ 16UL ];
+  uchar const * identity_key;
+
+  fd_wksp_t * in_mem;
+  ulong       in_chunk0;
+  ulong       in_wmark;
+} fd_gui_ctx_t;
+
+FD_FN_CONST static inline ulong
+scratch_align( void ) {
+  return 128UL;
+}
+
+FD_FN_PURE static inline ulong
+scratch_footprint( fd_topo_tile_t const * tile ) {
+  (void)tile;
+
+  ulong l = FD_LAYOUT_INIT;
+  l = FD_LAYOUT_APPEND( l, alignof( fd_gui_ctx_t ), sizeof( fd_gui_ctx_t ) );
+  l = FD_LAYOUT_APPEND( l, fd_http_server_align(),  fd_http_server_footprint( GUI_PARAMS ) );
+  l = FD_LAYOUT_APPEND( l, fd_gui_align(),          fd_gui_footprint() );
+  return FD_LAYOUT_FINI( l, scratch_align() );
+}
+
+FD_FN_CONST static inline void *
+mux_ctx( void * scratch ) {
+  return (void*)fd_ulong_align_up( (ulong)scratch, alignof( fd_gui_ctx_t ) );
+}
+
+static void
+before_credit( void *             _ctx,
+               fd_mux_context_t * mux ) {
+  (void)mux;
+
+  fd_gui_ctx_t * ctx = (fd_gui_ctx_t *)_ctx;
+
+  fd_http_server_poll( ctx->gui_server );
+  fd_gui_poll( ctx->gui );
+}
+
+static inline void
+during_frag( void * _ctx,
+             ulong  in_idx,
+             ulong  seq,
+             ulong  sig,
+             ulong  chunk,
+             ulong  sz,
+             int *  opt_filter ) {
+  (void)in_idx;
+  (void)seq;
+  (void)sig;
+  (void)opt_filter;
+
+  fd_gui_ctx_t * ctx = (fd_gui_ctx_t *)_ctx;
+
+  uchar * src = (uchar *)fd_chunk_to_laddr( ctx->in_mem, chunk );
+
+   /* ... todo... sigh, sz is not correct since it's too big */
+  if( sig==FD_PLUGIN_MSG_GOSSIP_UPDATE || sig==FD_PLUGIN_MSG_VOTE_ACCOUNT_UPDATE || sig==FD_PLUGIN_MSG_VALIDATOR_INFO ) sz = 8UL + 40200UL*(58UL+12UL*34UL);
+  else if( sig==FD_PLUGIN_MSG_LEADER_SCHEDULE ) sz = 40UL + 40200UL*40UL;
+
+  if( FD_UNLIKELY( chunk<ctx->in_chunk0 || chunk>ctx->in_wmark || sz>sizeof( ctx->buf ) ) )
+    FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz, ctx->in_chunk0, ctx->in_wmark ));
+
+  fd_memcpy( ctx->buf, src, sz );
+}
+
+static inline void
+after_frag( void *             _ctx,
+            ulong              in_idx,
+            ulong              seq,
+            ulong *            opt_sig,
+            ulong *            opt_chunk,
+            ulong *            opt_sz,
+            ulong *            opt_tsorig,
+            int *              opt_filter,
+            fd_mux_context_t * mux ) {
+  (void)in_idx;
+  (void)seq;
+  (void)opt_chunk;
+  (void)opt_tsorig;
+  (void)opt_filter;
+  (void)mux;
+
+  fd_gui_ctx_t * ctx = (fd_gui_ctx_t *)_ctx;
+
+  fd_gui_plugin_message( ctx->gui, *opt_sig, ctx->buf, *opt_sz );
```

Comment:
Is `*opt_sz` accurate here? From `during_frag` it seems like no. The value computed in `during_frag` isn't passed back to the mux.